### PR TITLE
crimson/os: rewrite ordering using std::strong_ordering

### PR DESCRIPTION
**Pull request by tchaikov**
crimson/os: rewrite ordering using std::strong_ordering

the goals are

1. to use std::strong_ordering mechinary for better readability
   and maintainbility
2. to remove unnecessary abstraction
3. use concept for better error message and readability.

changes:

* replace MatchKindCMP with std::strong_ordering
* replace compare_to() with operator<=>
* introduce a concept `IsFullKey` so we can use it when developing
  generic facilities to operate on both materialized key or view.
* use `IsFullKey` in place of `KeyT`.
* use type instead of `KeyT` when appropriate.
* use the constrained key type instead of `full_key_t<KT>`
  full_key_t<KT>`

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/crimson/os/seastore/onode_manager/staged-fltree/tree.h [hunk 1]
```diff
@@ -52,9 +52,9 @@ class Btree {
    *
    * This defines the internal order of Btree from ghobject_t perspective.
    */
-  static int compare(const ghobject_t &l, const ghobject_t &r) {
-    return static_cast<int>(key_hobj_t(l).compare_to(key_hobj_t(r)));
-  }
+  // static int compare(const ghobject_t &l, const ghobject_t &r) {
+  //   return static_cast<int>(key_hobj_t(l) <=> key_hobj_t(r));
+  // }
```
Comment (tchaikov) [inline src/crimson/os/seastore/onode_manager/staged-fltree/tree.h:57]: will extract this into a dedicated commit.

File: src/crimson/os/seastore/onode_manager/staged-fltree/tree.h [hunk 2]
```diff
@@ -104,12 +104,8 @@ class Btree {
...
     }
 
-    bool operator>(const Cursor& o) const { return (int)compare_to(o) > 0; }
-    bool operator>=(const Cursor& o) const { return (int)compare_to(o) >= 0; }
-    bool operator<(const Cursor& o) const { return (int)compare_to(o) < 0; }
-    bool operator<=(const Cursor& o) const { return (int)compare_to(o) <= 0; }
-    bool operator==(const Cursor& o) const { return (int)compare_to(o) == 0; }
-    bool operator!=(const Cursor& o) const { return (int)compare_to(o) != 0; }
+    bool operator==(const Cursor& o) const { return operator<=>(o) == 0; }
+    bool operator!=(const Cursor& o) const { return !operator==(o); }
```
Comment (tchaikov) [inline src/crimson/os/seastore/onode_manager/staged-fltree/tree.h:108]: this one can be dropped.

**Comment by tchaikov** [comment]
i will 

- address the comments 
- cleanup the commits
- fix the FTBFS of tests

after the first round of reviews.

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
```
0.	Program arguments: /usr/bin/clang++-14 -g -fPIC -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free -Wall -fno-strict-aliasing -fsigned-char -Wtype-limits -Wignored-qualifiers -Wpointer-arith -Werror=format-security -Winit-self -Wno-unknown-pragmas -Wnon-virtual-dtor -Wno-ignored-qualifiers -ftemplate-depth-1024 -Wpessimizing-move -Wredundant-move -Wno-inconsistent-missing-override -Wno-mismatched-tags -Wno-unused-private-field -Wno-address-of-packed-member -Wno-unused-function -Wno-unused-local-typedef -Wno-varargs -Wno-gnu-designator -Wno-missing-braces -Wno-parentheses -Wno-deprecated-register -fdiagnostics-color=auto -ftemplate-backtrace-limit=0 -Wno-non-virtual-dtor -std=gnu++17 -Wno-error=unused-result -Wno-error=#warnings -fstack-clash-protection -fsanitize=address -fsanitize=undefined -fno-sanitize=vptr -std=c++2a -Werror -DBOOST_ASIO_DISABLE_CONCEPTS -DBOOST_ASIO_DISABLE_THREAD_KEYWORD_EXTENSION -DBOOST_ASIO_USE_TS_EXECUTOR_AS_DEFAULT -DHAVE_CONFIG_H -DSEASTAR_API_LEVEL=6 -DSEASTAR_DEBUG -DSEASTAR_DEBUG_SHARED_PTR -DSEASTAR_DEFAULT_ALLOCATOR -DSEASTAR_SCHEDULING_GROUPS_COUNT=16 -DSEASTAR_SHUFFLE_TASK_QUEUE -DSEASTAR_TYPE_ERASE_MORE -DWITH_SEASTAR=1 -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_REENTRANT -D_THREAD_SAFE -D__CEPH__ -D__STDC_FORMAT_MACROS -D__linux__ -Isrc/include -I../src -I../src/seastar/include -Isrc/seastar/gen/include -isystem boost/include -isystem include -isystem ../src/xxHash -isystem ../src/rapidjson/include -isystem ../src/fmt/include -isystem ../src/jaegertracing/opentelemetry-cpp/api/include -isystem ../src/jaegertracing/opentelemetry-cpp/exporters/jaeger/include -isystem ../src/jaegertracing/opentelemetry-cpp/ext/include -isystem ../src/jaegertracing/opentelemetry-cpp/sdk/include -DCEPH_DEBUG_MUTEX -D_GLIBCXX_ASSERTIONS -U_FORTIFY_SOURCE -DSEASTAR_SSTRING -c -o src/crimson/os/seastore/CMakeFiles/crimson-seastore.dir/onode_manager/staged-fltree/stages/node_stage.cc.o ../src/crimson/os/seastore/onode_manager/staged-fltree/stages/node_stage.cc
1.	../src/crimson/os/seastore/onode_manager/staged-fltree/stages/node_stage.cc:403:68: current parser token ';'
2.	../src/crimson/os/seastore/onode_manager/staged-fltree/stages/node_stage.cc:9:1: parsing namespace 'crimson'
3.	../src/crimson/os/seastore/onode_manager/staged-fltree/stages/node_stage.h:202:3: instantiating function definition 'crimson::os::seastore::onode::node_extent_t<crimson::os::seastore::onode::_node_fields_013_t<crimson::os::seastore::onode::_slot_t<crimson::os::seastore::onode::shard_pool_crush_t, crimson::os::seastore::onode::field_type_t::N0>>, crimson::os::seastore::onode::node_type_t::INTERNAL>::Appender<crimson::os::seastore::onode::key_view_t>::Appender'
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
/usr/lib/x86_64-linux-gnu/libLLVM-14.so.1(_ZN4llvm3sys15PrintStackTraceERNS_11raw_ostreamEi+0x31)[0x7f1676b65311]
/usr/lib/x86_64-linux-gnu/libLLVM-14.so.1(_ZN4llvm3sys17RunSignalHandlersEv+0xee)[0x7f1676b6305e]
/usr/lib/x86_64-linux-gnu/libLLVM-14.so.1(_ZN4llvm3sys15CleanupOnSignalEm+0x100)[0x7f1676b646d0]
/usr/lib/x86_64-linux-gnu/libLLVM-14.so.1(+0xd77a6f)[0x7f1676a91a6f]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x153c0)[0x7f167fdd73c0]
/usr/lib/x86_64-linux-gnu/libclang-cpp.so.14(+0xd167c1)[0x7f167d2f07c1]
/usr/lib/x86_64-linux-gnu/libclang-cpp.so.14(+0xd125e8)[0x7f167d2ec5e8]
/usr/lib/x86_64-linux-gnu/libclang-cpp.so.14(+0xceb6ed)[0x7f167d2c56ed]
/usr/lib/x86_64-linux-gnu/libclang-cpp.so.14(+0xce75f4)[0x7f167d2c15f4]
```

**Review by cyx1231st — APPROVED** [review]
